multi-s3-output/out_multi_s3.go

```diff
@@ -0,0 +1,333 @@
+		return output.FLB_ERROR
+	}
+
+	ctx := NewPluginContext(plugin, clientSet)
+
+	output.FLBPluginSetContext(plugin, ctx)
+
+	return output.FLB_OK
+}
+
+//export FLBPluginFlush
+func FLBPluginFlush(data unsafe.Pointer, length C.int, tag *C.char) int {
+	klog.Warning("[multi-s3] Flush called for unknown instance")
+	return output.FLB_OK
+}
+
+//export FLBPluginFlushCtx
+func FLBPluginFlushCtx(ctx, data unsafe.Pointer, length C.int, cTag *C.char) int {
+	// Type assert context back into the original type for the Go variable
+	tag := C.GoString(cTag)
+	pluginCtx := output.FLBPluginGetContext(ctx).(*PluginContext)
+	kubeClient := pluginCtx.ClientSet
+	s := strings.Split(tag, ".")
+	namespace := s[1]
+	podName := s[2]
+	klog.Infof("[multi-s3] Flush called for namespace: %q, pod: %q", namespace, podName)
+
+	pod, err := kubeClient.CoreV1().Pods(namespace).Get(context.TODO(), podName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Errorf("[multi-s3] Pod %q not found (%s). Skipping logs...", podName, err)
+			return output.FLB_ERROR // no reason to retry
+		} else {
+			klog.Warningf("[multi-s3] Unable to retrieve Pod %q metadata: %s.", podName, err)
+			return output.FLB_RETRY
+		}
+	}
+
+	secretName := pod.Annotations[pluginCtx.ArtifactSecretAnnotation]
+	if secretName == "" {
+		klog.Warningf("[multi-s3] The secret name for artifact repository is not set. Using default...")
+		secretName = "mlpipeline-minio-artifact"
+	}
+	klog.Infof("[multi-s3] Secret: %q", secretName)
+
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Errorf("[multi-s3] Secret %q not found (%s). Skipping logs...", secretName, err)
+			return output.FLB_ERROR // no reason to retry
+		} else {
+			klog.Infof("[multi-s3] Warning! Unable to retrieve secret %q: %s.", secretName, err)
+			return output.FLB_RETRY
+		}
+	}
+
+	pipelineRun := pod.Labels[pluginCtx.PipelineRunLabel]
+	pipelineTask := pod.Labels[pluginCtx.PipelineTaskLabel]
+	pipelineTaskRunId := pod.Labels[pluginCtx.PipelineTaskRunLabel]
+
+	accessKey, ok := secret.Data["accesskey"]
+	if !ok {
+		klog.Errorf("[multi-s3] Invalid s3 secret %q. No access key.", secretName)
+		return output.FLB_ERROR // no reason to retry
+	}
+	secretKey, ok := secret.Data["secretkey"]
+	if !ok {
+		klog.Errorf("[multi-s3] Invalid s3 secret %q. No secret key.", secretName)
+		return output.FLB_ERROR // no reason to retry
+	}
+
+	s3ctx := S3BucketContext{
+		Endpoint:       pod.Annotations[pluginCtx.ArtifactEndpointAnnotation],
+		EndpointSchema: pod.Annotations[pluginCtx.ArtifactEndpointSchemeAnnotation],
+		AccessKey:      string(accessKey),
+		SecretKey:      string(secretKey),
+		Bucket:         pod.Annotations[pluginCtx.ArtifactBucketAnnotation],
+	}
+
+	var buf bytes.Buffer
+	b := C.GoBytes(data, C.int(length))
+	result := ArchiveLog(b, &buf)
+	if result != output.FLB_OK {
+		return result
+	}
```

Comment:
1. Depends on what you consider "beauty". `bytes.Buffer`'s `io.Reader` aspect gets invalidated on every write (because reallocations are possible), so it must be handled with extreme care, while `[]byte`'s readers are safe (of course converting `bytes.Buffer` into an `io.Reader` and discarding the original `bytes.Buffer` isn't half-bad either). I'd say "safer" code, i.e. code with less possible "gotchas", is cleaner. If that rule didn't exist, nobody would use `string`, considering `[]byte` has pretty much the same possibilities + mutation, and these two types have lots of functions, which doubles the code --- there are good reasons for standard library to be doubling so much code. 

2. It's not conversion-back-and-forth. `func (b *Buffer) Bytes() []byte` isn't a conversion, it's a getter.